File: engine-reference/forge/lands.py
"""Clasificador de tierras (referencia canónica del motor Forge).

Espejo 1:1 de `forge_engine/lib/src/lands.dart`. Shocks/checks/fetches y
demás no-básicas de la colección entran en la manabase por función real,
no solo básicas proporcionales.
"""
from __future__ import annotations
import re
from dataclasses import dataclass, field

BASIC_TYPE_COLOR = {
    "Plains": "W",
    "Island": "U",
    "Swamp": "B",
    "Mountain": "R",
    "Forest": "G",
}

BASIC_NAMES = {
    "Plains", "Island", "Swamp", "Mountain", "Forest",
    "Snow-Covered Plains", "Snow-Covered Island", "Snow-Covered Swamp",
    "Snow-Covered Mountain", "Snow-Covered Forest",
}

ADD_CLAUSE = re.compile(r"add ([^\n.]*)")
SYMBOL = re.compile(r"\{([wubrgc])\}")
FETCH = re.compile(r"search your library for ([^\n.]*land[^\n.]*)")
TAPPED = re.compile(r"enters (the battlefield )?tapped")


def _line_at(text: str, pos: int) -> str:
    """Línea (delimitada por '\\n') que contiene la posición pos."""
    start = text.rfind("\n", 0, pos + 1) + 1
    nl = text.find("\n", pos)
    end = len(text) if nl == -1 else nl
    return text[start:end]


@dataclass
class LandProfile:
    produces: set[str]
    tapped: str  # "never" | "conditional" | "always"
    is_fetch: bool
    fetches: set[str] = field(default_factory=set)
    is_basic: bool = False

    @property
    def is_utility(self) -> bool:
        return not self.produces and not self.is_fetch


def land_profile(card: dict) -> LandProfile:
    text = (card.get("oracle") or "").lower()

    produces: set[str] = set()
    for sub in card.get("subtypes", []):
        color = BASIC_TYPE_COLOR.get(sub)
        if color:
            produces.add(color)
    for m in ADD_CLAUSE.finditer(text):
        clause = m.group(1)
        if "one mana of any color" in clause or "mana of any one color" in clause:
            produces.update("WUBRG")
        for s in SYMBOL.finditer(clause):
            ch = s.group(1).upper()
            if ch != "C":
                produces.add(ch)

    tapped = "never"
    tapped_match = TAPPED.search(text)
    if tapped_match:
        line = _line_at(text, tapped_match.start())
        tapped = "conditional" if ("unless" in line or " if " in line) else "always"

    is_fetch = False
    fetches: set[str] = set()
    fetch_match = FETCH.search(text)
    if fetch_match:
        is_fetch = True
        clause = fetch_match.group(1)
        for name, color in BASIC_TYPE_COLOR.items():
            if name.lower() in clause:
                fetches.add(color)

    is_basic = "Basic" in card.get("types", []) or card.get("name") in BASIC_NAMES

    return LandProfile(
        produces=produces,
        tapped=tapped,
        is_fetch=is_fetch,
        fetches=fetches,
        is_basic=is_basic,
    )


def source_of(profile: LandProfile, color: str, deck_colors: str) -> bool:
    """¿Cuenta esta tierra como fuente de color en un mazo de deck_colors?"""
    if color in profile.produces:
        return True
    if profile.is_fetch:
        return color in deck_colors if not profile.fetches else color in profile.fetches
    return False

File: forge_engine/lib/src/lands.dart
import 'models.dart';

/// Clasificador de tierras — espejo 1:1 de `engine-reference/forge/lands.py`.
///
/// Fase 1 (maná): shocks/checks/fetches y demás no-básicas de la colección
/// entran en la manabase por función real, no solo básicas proporcionales.

enum TappedKind { never, conditional, always }

const Map<String, String> _basicTypeColor = {
  'Plains': 'W',
  'Island': 'U',
  'Swamp': 'B',
  'Mountain': 'R',
  'Forest': 'G',
};

const Set<String> _basicNames = {
  'Plains', 'Island', 'Swamp', 'Mountain', 'Forest', //
  'Snow-Covered Plains', 'Snow-Covered Island', 'Snow-Covered Swamp',
  'Snow-Covered Mountain', 'Snow-Covered Forest',
};

final RegExp _addClause = RegExp(r'add ([^\n.]*)');
final RegExp _sym = RegExp(r'\{([wubrgc])\}');
final RegExp _fetch = RegExp(r'search your library for ([^\n.]*land[^\n.]*)');
final RegExp _tappedRe = RegExp(r'enters (the battlefield )?tapped');

/// Línea del texto (delimitada por '\n') que contiene la posición [pos].
String _lineAt(String text, int pos) {
  final start = text.lastIndexOf('\n', pos) + 1;
  final nl = text.indexOf('\n', pos);
  final end = nl == -1 ? text.length : nl;
  return text.substring(start, end);
}

class LandProfile {
  final Set<String> produces; // WUBRG que puede dar (duales tipadas incluidas)
  final TappedKind tapped;
  final bool isFetch;
  final Set<String> fetches; // colores buscables; vacío + isFetch = básica genérica
  final bool isBasic;

  const LandProfile({
    required this.produces,
    required this.tapped,
    required this.isFetch,
    required this.fetches,
    required this.isBasic,
  });

  bool get isUtility => produces.isEmpty && !isFetch;

  /// ¿Cuenta como fuente de [color] en un mazo de [deckColors]?
  bool sourceOf(String color, String deckColors) =>
      produces.contains(color) ||
      (isFetch &&
          (fetches.isEmpty
              ? deckColors.contains(color)
              : fetches.contains(color)));

  static LandProfile fromCard(Card card) {
    final text = card.oracle.toLowerCase();

    final produces = <String>{};
    for (final sub in card.subtypes) {
      final c = _basicTypeColor[sub];
      if (c != null) produces.add(c);
    }
    for (final m in _addClause.allMatches(text)) {
      final clause = m.group(1)!;
      if (clause.contains('one mana of any color') ||
          clause.contains('mana of any one color')) {
        produces.addAll(const ['W', 'U', 'B', 'R', 'G']);
      }
      for (final s in _sym.allMatches(clause)) {
        final ch = s.group(1)!.toUpperCase();
        if (ch != 'C') produces.add(ch);
      }
    }

    var tapped = TappedKind.never;
    final tappedMatch = _tappedRe.firstMatch(text);
    if (tappedMatch != null) {
      final line = _lineAt(text, tappedMatch.start);
      tapped = (line.contains('unless') || line.contains(' if '))
          ? TappedKind.conditional
          : TappedKind.always;
    }

    var isFetch = false;
    final fetches = <String>{};
    final fetchMatch = _fetch.firstMatch(text);
    if (fetchMatch != null) {
      isFetch = true;
      final clause = fetchMatch.group(1)!;
      for (final entry in _basicTypeColor.entries) {
        if (clause.contains(entry.key.toLowerCase())) fetches.add(entry.value);
      }
    }

    final isBasic =
        card.types.contains('Basic') || _basicNames.contains(card.name);

    // Básicas "de cortesía" de la app: sin oracle ni subtipos. El color
    // sale del nombre, o la manabase entera se queda sin candidatas.
    if (produces.isEmpty && isBasic) {
      final c = _basicTypeColor[card.name.replaceFirst('Snow-Covered ', '')];
      if (c != null) produces.add(c);
    }

    return LandProfile(
      produces: produces,
      tapped: tapped,
      isFetch: isFetch,
      fetches: fetches,
      isBasic: isBasic,
    );
  }
}
